DatadogSessionReplay/Sources/Recorder/RecordingCoordinator.swift

```diff
@@ -43,62 +51,62 @@ internal class RecordingCoordinator {
         telemetry: Telemetry,
-        startRecordingImmediately: Bool,
-        methodCallTelemetrySamplingRate: Float = 0.1
-    ) {
+        recordingTrigger: RecordingTriggering,
+        methodCallTelemetrySamplingRate: Float = 0.1,
+        dateProvider: DateProvider = SystemDateProvider(),
+        queue: Queue = MainAsyncQueue()
+    ) throws {
         self.recorder = recorder
-        self.scheduler = scheduler
         self.sampler = sampler
         self.textAndInputPrivacy = textAndInputPrivacy
         self.imagePrivacy = imagePrivacy
         self.touchPrivacy = touchPrivacy
         self.srContextPublisher = srContextPublisher
         self.telemetry = telemetry
+        self.recordingTrigger = recordingTrigger
         self.methodCallTelemetrySamplingRate = methodCallTelemetrySamplingRate
+        self.dateProvider = dateProvider
+        self.queue = queue
 
         srContextPublisher.setHasReplay(false)
 
-        scheduler.schedule { [weak self] in self?.captureNextRecord() }
-
-        // Start recording immediately if specified.
-        if startRecordingImmediately {
-            startRecording()
+        // Observe changes in the RUM context on the main thread.
+        rumContextObserver.observe(on: queue) { [weak self] in
+            self?.onRUMContextChanged(rumContext: $0)
         }
-
-        // Observe changes in the RUM context.
-        rumContextObserver.observe(on: scheduler.queue) { [weak self] in self?.onRUMContextChanged(rumContext: $0) }
     }
 
     /// Enables recording based on user request.
     func startRecording() {
-        scheduler.queue.run { [weak self] in
+        queue.run { [weak self] in
             self?.recordingEnabled = true
             self?.evaluateRecordingConditions()
         }
     }
 
     /// Disables recording based on user request.
     func stopRecording() {
-        scheduler.queue.run { [weak self] in
+        queue.run { [weak self] in
             self?.recordingEnabled = false
             self?.evaluateRecordingConditions()
         }
     }
 
-    // MARK: Private
-
-    /// Evaluates whether recording should start or stop based on user request and sampling.
     private func evaluateRecordingConditions() {
-       if recordingEnabled && isSampled {
-           scheduler.start()
-       } else {
-           scheduler.stop()
-       }
-       updateHasReplay()
-   }
+        if recordingEnabled && isSampled && currentRUMContext != nil {
+            recordingTrigger.startWatchingTriggers { [weak self] in
+                self?.didTrigger()
+            }
+        } else {
+            recordingTrigger.stopWatchingTriggers()
+        }
+        updateHasReplay()
+    }
+
+    // MARK: Private
 
     private func onRUMContextChanged(rumContext: RUMContext?) {
-        if currentRUMContext?.sessionID != rumContext?.sessionID || currentRUMContext == nil {
+        if currentRUMContext?.sessionID != rumContext?.sessionID || currentRUMContext == nil && rumContext != nil {
```

Comment:
I'm not sure I understand this addition. In which scenario `currentRUMContext` would be `nil` and `rumContext` **not** `nil`? When switching views? Is it really what we want to check here? 

Parenthesis could improve readability.

Reply:
Yeah - it's actually unnecessary, because we have checks in other places.

Normally this scenario is impossible, maybe except for some edge cases where RUM gets deactivated during the runtime. It's here mostly to satisfy unit tests.